typing/typecore.ml

```diff
@@ -1006,34 +1006,43 @@ let check_scope_escape loc env level ty =
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
-(* type_pat propagates the expected type as well as maps for
-   constructors and labels.
-   Unification may update the typing environment. *)
-(* constrs <> None => called from parmatch: backtrack on or-patterns
-   explode > 0 => explode Ppat_any for gadts *)
-let rec type_pat ?(exception_allowed=false) ~constrs ~labels ~no_existentials
-          ~mode ~explode ~env sp expected_ty k =
+type pattern_typing_data = {
+  exception_allowed: bool;
+
+  (* maps for constructors and labels *)
+  (* constrs <> None => called from parmatch: backtrack on or-patterns *)
+  constrs: (string, Types.constructor_description) Hashtbl.t option;
+  labels: (string, Types.label_description) Hashtbl.t option;
+
+  no_existentials: existential_restriction option;
+  mode: type_pat_mode;
+
+  (* explode > 0 => explode Ppat_any for gadts *)
+  explode: int;
+}
+
+let rec type_pat data ~env sp expected_ty k =
   Builtin_attributes.warning_scope sp.ppat_attributes
     (fun () ->
-       type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
-         ~explode ~env sp expected_ty k
+       type_pat_aux data ~env sp expected_ty k
     )
 
-and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
-      ~explode ~env sp expected_ty k =
-  let mode' = if mode = Splitting_or then Normal else mode in
-  let type_pat ?(exception_allowed=false) ?(constrs=constrs) ?(labels=labels)
-        ?(mode=mode') ?(explode=explode) ?(env=env) =
-    type_pat ~exception_allowed ~constrs ~labels ~no_existentials ~mode ~explode
-      ~env
-  in
+and type_pat_aux data ~env sp expected_ty k =
+  let data =
+    let mode = if data.mode = Splitting_or then Normal else data.mode in
+    { data with mode } in
+  let type_pat ?(exception_allowed=false) data ?(env=env) =
```

Comment:
I find the the name `exception_allowed` potentially confusing on call sites where
```OCaml
type_pat ~exception_allowed:true data 
```
only means that exceptions are allowed if they were already allowed in `data`.
Maybe a label like `exception_mask` would better convey that we are keeping the exception permission rather than reclaiming it?

Reply:
I thought about this as well. The way I think of it, the callsite either allows exception or it does not, this is what the parameter says, and then the global choice depends on both the callsite and the global context.

I'm not sure it's worth changing things here, given that this `exception_allowed` part disappears with #8970. If we want to do a change here, I would rather take it out of the `data` record and make it explicit at all callsites.

Reply:
Ok, the name indeed seems coherent with the idea that it applies to the call site specifically.